Quarkus fails to start up with composite Gradle build when using Kotlin (Java works correctly)

### Describe the bug

When both the root gradle project and the gradle subproject in a composite gradle build use Quarkus with Kotlin, the project throws a runtime exception and fails to start up.

### Expected behavior

The project starts up successfully.

### Actual behavior

The following exception is thrown at runtime and the project fails to start up:

```
2025-04-17 14:01:57,148 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: jakarta.enterprise.inject.spi.DeploymentException: jakarta.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type org.example.SomeBean and qualifiers [@Default]
        - injection target: parameter 'someBean' of org.example.ExampleResource constructor
        - declared on CLASS bean [types=[org.example.ExampleResource, java.lang.Object], qualifiers=[@Default, @Any], target=org.example.ExampleResource]
        at io.quarkus.arc.processor.BeanDeployment.processErrors(BeanDeployment.java:1581)
        at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:338)
        at io.quarkus.arc.processor.BeanProcessor.initialize(BeanProcessor.java:178)
        at io.quarkus.arc.deployment.ArcProcessor.validate(ArcProcessor.java:489)
        at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:255)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
        at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
        at java.base/java.lang.Thread.run(Thread.java:1583)
        at org.jboss.threads.JBossThread.run(JBossThread.java:499)
Caused by: jakarta.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type org.example.SomeBean and qualifiers [@Default]
        - injection target: parameter 'someBean' of org.example.ExampleResource constructor
        - declared on CLASS bean [types=[org.example.ExampleResource, java.lang.Object], qualifiers=[@Default, @Any], target=org.example.ExampleResource]
        at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:547)
        at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:689)
        at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:323)
        ... 12 more
```


If the root project doesn't inject any beans from the dependency and instead only accesses a plain Kotlin class from the dependency, then this exception is thrown instead and the project still fails to start up:

```
2025-04-17 14:12:09,066 WARN  [io.qua.dep.ste.ReflectiveHierarchyStep] (build-4) Unable to properly register the hierarchy of the following classes for reflection as they are not in the Jandex index:
     - org.example.SomeDataClass (source: ResteasyReactiveProcessor > org.example.ExampleResource[org.example.SomeDataClass hello()] > org.example.SomeDataClass > org.example.SomeDataClass)
Consider adding them to the index either by creating a Jandex index for your dependency via the Maven plugin, an empty META-INF/beans.xml or quarkus.index-dependency properties.
2025-04-17 14:12:09,126 ERROR [io.qua.run.Quarkus] (Quarkus Main Thread) Error running Quarkus: java.lang.ExceptionInInitializerError
        at java.base/jdk.internal.misc.Unsafe.allocateInstance(Native Method)
        at java.base/java.lang.invoke.DirectMethodHandle.allocateInstance(DirectMethodHandle.java:501)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:79)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:51)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:144)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:116)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
        ... 10 more
Caused by: java.lang.RuntimeException: java.lang.ClassNotFoundException: org.example.SomeDataClass
        at org.jboss.resteasy.reactive.common.util.types.TypeSignatureParser.loadClass(TypeSignatureParser.java:165)
        at org.jboss.resteasy.reactive.common.util.types.TypeSignatureParser.parseReference(TypeSignatureParser.java:146)
        at org.jboss.resteasy.reactive.common.util.types.TypeSignatureParser.parseType(TypeSignatureParser.java:54)
        at org.jboss.resteasy.reactive.common.util.types.TypeSignatureParser.parse(TypeSignatureParser.java:170)
        at org.jboss.resteasy.reactive.server.core.startup.RuntimeResourceDeployment.buildResourceMethod(RuntimeResourceDeployment.java:389)
        at org.jboss.resteasy.reactive.server.core.startup.RuntimeDeploymentManager.deploy(RuntimeDeploymentManager.java:137)
        at io.quarkus.resteasy.reactive.server.runtime.ResteasyReactiveRecorder.createDeployment(ResteasyReactiveRecorder.java:161)
        at io.quarkus.runner.recorded.ResteasyReactiveProcessor$setupDeployment944680882.deploy_0(Unknown Source)
        at io.quarkus.runner.recorded.ResteasyReactiveProcessor$setupDeployment944680882.deploy(Unknown Source)
        ... 11 more
Caused by: java.lang.ClassNotFoundException: org.example.SomeDataClass
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:569)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:517)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:569)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:517)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:534)
        at java.base/java.lang.Class.forName(Class.java:513)
        at org.jboss.resteasy.reactive.common.util.types.TypeSignatureParser.loadClass(TypeSignatureParser.java:163)
        ... 19 more
```

### How to Reproduce?

Reproducer: https://github.com/lspahija/quarkus-complex-build-bug-reproducer

There are two branches: `not-working-kotlin` (default branch) and `works-java`

Steps to reproduce the behavior:
1. git clone https://github.com/lspahija/quarkus-complex-build-bug-reproducer
2. cd quarkus-complex-build-bug-reproducer
3. ./gradlew quarkusDev
4. see exception in the console and note that the project has failed to start up

If you wish to see the equivalent java project working successfully:
1. git checkout works-java
2. ./gradlew quarkusDev
3. note that everything works fine

### Output of `uname -a` or `ver`

Darwin MacBookPro 24.3.0 Darwin Kernel Version 24.3.0: Thu Jan  2 20:24:16 PST 2025; root:xnu-11215.81.4~3/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

openjdk version "21.0.6" 2025-01-21 LTS OpenJDK Runtime Environment Zulu21.40+17-CA (build 21.0.6+7-LTS) OpenJDK 64-Bit Server VM Zulu21.40+17-CA (build 21.0.6+7-LTS, mixed mode, sharing)

### Quarkus version or git rev

3.21.3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 8.12

### Additional information

_No response_

--- a/devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/tooling/GradleApplicationModelBuilder.java
+++ b/devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/tooling/GradleApplicationModelBuilder.java
@@ -17,6 +17,7 @@ import java.util.Map;
 import java.util.Objects;
 import java.util.Properties;
 import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
 
 import org.gradle.api.Project;
 import org.gradle.api.Task;
@@ -36,7 +37,7 @@ import org.gradle.api.tasks.testing.Test;
 import org.gradle.internal.composite.IncludedBuildInternal;
 import org.gradle.language.jvm.tasks.ProcessResources;
 import org.gradle.tooling.provider.model.ParameterizedToolingModelBuilder;
-import org.jetbrains.kotlin.gradle.tasks.KotlinJvmCompile;
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompileTool;
 
 import io.quarkus.bootstrap.BootstrapConstants;
 import io.quarkus.bootstrap.model.ApplicationModel;
@@ -60,7 +61,6 @@ import io.quarkus.maven.dependency.DependencyFlags;
 import io.quarkus.maven.dependency.GACT;
 import io.quarkus.maven.dependency.GACTV;
 import io.quarkus.maven.dependency.GAV;
-import io.quarkus.maven.dependency.ResolvedDependency;
 import io.quarkus.maven.dependency.ResolvedDependencyBuilder;
 import io.quarkus.paths.PathCollection;
 import io.quarkus.paths.PathList;
@@ -352,23 +352,20 @@ public class GradleApplicationModelBuilder implements ParameterizedToolingModelB
             }
 
             PathCollection paths = null;
-            if (workspaceDiscovery && a.getId().getComponentIdentifier() instanceof ProjectComponentIdentifier) {
-
-                Project projectDep = project.getRootProject().findProject(
-                        ((ProjectComponentIdentifier) a.getId().getComponentIdentifier()).getProjectPath());
+            if (workspaceDiscovery && a.getId().getComponentIdentifier() instanceof ProjectComponentIdentifier compId) {
+                Project projectDep = project.getRootProject().findProject(compId.getProjectPath());
                 SourceSetContainer sourceSets = projectDep == null ? null
                         : projectDep.getExtensions().findByType(SourceSetContainer.class);
 
                 final String classifier = a.getClassifier();
                 if (classifier == null || classifier.isEmpty()) {
                     final IncludedBuild includedBuild = ToolingUtils.includedBuild(project.getRootProject(),
-                            ((ProjectComponentIdentifier) a.getId().getComponentIdentifier()).getBuild().getName());
+                            compId.getBuild().getName());
                     if (includedBuild != null) {
                         final PathList.Builder pathBuilder = PathList.builder();
 
-                        if (includedBuild instanceof IncludedBuildInternal) {
-                            projectDep = ToolingUtils.includedBuildProject((IncludedBuildInternal) includedBuild,
-                                    ((ProjectComponentIdentifier) a.getId().getComponentIdentifier()).getProjectPath());
+                        if (includedBuild instanceof IncludedBuildInternal ib) {
+                            projectDep = ToolingUtils.includedBuildProject(ib, compId.getProjectPath());
                         }
                         if (projectDep != null) {
                             projectModule = initProjectModuleAndBuildPaths(projectDep, a, modelBuilder, depBuilder,
@@ -517,7 +514,6 @@ public class GradleApplicationModelBuilder implements ParameterizedToolingModelB
 
     private static void initProjectModule(Project project, WorkspaceModule.Mutable module, SourceSet sourceSet,
             String classifier) {
-
         if (sourceSet == null) {
             return;
         }
@@ -570,12 +566,30 @@ public class GradleApplicationModelBuilder implements ParameterizedToolingModelB
 
     private static void maybeConfigureKotlinJvmCompile(Project project, FileCollection allClassesDirs,
             List<SourceDir> sourceDirs, SourceSet sourceSet) {
-        // This "try/catch" is needed because of the way the "quarkus-cli" Gradle tests work. Without it, the tests fail.
-        try {
-            Class.forName("org.jetbrains.kotlin.gradle.tasks.KotlinJvmCompile");
-            doConfigureKotlinJvmCompile(project, allClassesDirs, sourceDirs, sourceSet);
-        } catch (ClassNotFoundException e) {
-            // ignore
+        for (var task : project.getTasks()) {
+            if (task.getName().contains("compileKotlin") && task.getEnabled()) {
+                int originalSourceDirsSize = sourceDirs.size();
+
+                // This "try/catch" is needed because of the way the "quarkus-cli" Gradle tests work. Without it, the tests fail.
+                try {
+                    Class.forName("org.jetbrains.kotlin.gradle.tasks.KotlinCompileTool");
+                    doConfigureKotlinJvmCompile(project, allClassesDirs, sourceDirs, sourceSet);
+                } catch (ClassNotFoundException e) {
+                    // ignore
+                }
+                // if the above failed, there could still be a KotlinCompile task that's not easily discoverable
+                if (originalSourceDirsSize == sourceDirs.size()) {
+                    final Path outputDir = getClassesOutputDir(task);
+                    if (outputDir != null && task.getInputs().getHasInputs()) {
+                        task.getInputs().getSourceFiles().getAsFileTree().visit(visitor -> {
+                            if (visitor.getRelativePath().getSegments().length == 1) {
+                                sourceDirs.add(SourceDir.of(visitor.getFile().getParentFile().toPath(), outputDir));
+                            }
+                        });
+                    }
+                    break;
+                }
+            }
         }
     }
 
@@ -583,32 +597,26 @@ public class GradleApplicationModelBuilder implements ParameterizedToolingModelB
             List<SourceDir> sourceDirs, SourceSet sourceSet) {
         // Use KotlinJvmCompile.class in a separate method to prevent that maybeConfigureKotlinJvmCompile() runs into
         // a ClassNotFoundException due to actually using KotlinJvmCompile.class.
-        project.getTasks().withType(KotlinJvmCompile.class, t -> configureCompileTask(t.getSources().getAsFileTree(),
+        project.getTasks().withType(KotlinCompileTool.class, t -> configureCompileTask(t.getSources().getAsFileTree(),
                 t.getDestinationDirectory(), allClassesDirs, sourceDirs, t, sourceSet));
     }
 
     private static void configureCompileTask(FileTree sources, DirectoryProperty destinationDirectory,
             FileCollection allClassesDirs, List<SourceDir> sourceDirs, Task task, SourceSet sourceSet) {
-        if (!task.getEnabled()) {
+        if (!task.getEnabled() || sources.isEmpty()) {
             return;
         }
-        if (sources.isEmpty()) {
-            return;
-        }
-
         final File destDir = destinationDirectory.getAsFile().get();
         if (!allClassesDirs.contains(destDir)) {
             return;
         }
-        sources.visit(a -> {
+        sources.visit(visitor -> {
             // we are looking for the root dirs containing sources
-            if (a.getRelativePath().getSegments().length == 1) {
-                final File srcDir = a.getFile().getParentFile();
-
-                sourceDirs
-                        .add(new DefaultSourceDir(srcDir.toPath(), destDir.toPath(),
-                                findGeneratedSourceDir(destDir, sourceSet),
-                                Map.of("compiler", task.getName())));
+            if (visitor.getRelativePath().getSegments().length == 1) {
+                final File srcDir = visitor.getFile().getParentFile();
+                sourceDirs.add(new DefaultSourceDir(srcDir.toPath(), destDir.toPath(),
+                        findGeneratedSourceDir(destDir, sourceSet),
+                        Map.of("compiler", task.getName())));
             }
         });
     }
@@ -620,9 +628,6 @@ public class GradleApplicationModelBuilder implements ParameterizedToolingModelB
         }
         String language = destDir.getParentFile().getName(); // java
         String sourceSetName = destDir.getName(); // main
-        if (language == null) {
-            return null;
-        }
         // find the corresponding generated sources, same pattern, but under build/generated/sources/annotationProcessor/java/main
         for (File generatedDir : sourceSet.getOutput().getGeneratedSourcesDirs().getFiles()) {
             if (generatedDir.getParentFile() == null) {
@@ -632,6 +637,37 @@ public class GradleApplicationModelBuilder implements ParameterizedToolingModelB
                     && generatedDir.getParentFile().getName().equals(language)) {
                 return generatedDir.toPath();
             }
+        }
+        return null;
+    }
+
+    /**
+     * This method is meant to figure out the output directory containing class files for a compile task
+     * which is not available in the plugin classpath. An example would be KotlinCompile.
+     *
+     * @param compileTask a compile task
+     */
+    private static Path getClassesOutputDir(Task compileTask) {
+        if (compileTask.getOutputs().getHasOutput()) {
+            final AtomicReference<Path> result = new AtomicReference<>();
+            compileTask.getOutputs().getFiles().getAsFileTree().visit(visitor -> {
+                // We are looking for the first class file, since a compile task would typically
+                // have a single output location for classes.
+                // There in fact could be a few output locations, the rest though would typically be some internal caching bits
+                if (visitor.getName().endsWith(".class")) {
+                    visitor.stopVisiting();
+                    var file = visitor.getFile();
+                    int relativeSegments = visitor.getRelativePath().getSegments().length;
+                    while (file != null && relativeSegments > 0) {
+                        relativeSegments--;
+                        file = file.getParentFile();
+                    }
+                    if (file != null) {
+                        result.set(file.toPath());
+                    }
+                }
+            });
+            return result.get();
         }
         return null;
     }

--- a/integration-tests/gradle/src/main/resources/included-kotlin-build/acme-subproject-nested/acme-nested/build.gradle.kts
+++ b/integration-tests/gradle/src/main/resources/included-kotlin-build/acme-subproject-nested/acme-nested/build.gradle.kts
@@ -1,0 +1,48 @@
+plugins {
+    kotlin("jvm")
+    kotlin("plugin.allopen")
+    id("io.quarkus")
+}
+
+repositories {
+    mavenLocal {
+        content {
+            includeGroupByRegex("io.quarkus.*")
+            includeGroup("org.hibernate.orm")
+        }
+    }
+    mavenCentral()
+}
+
+val quarkusPlatformGroupId: String by project
+val quarkusPlatformArtifactId: String by project
+val quarkusPlatformVersion: String by project
+
+dependencies {
+    implementation(enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}"))
+    implementation("io.quarkus:quarkus-kotlin")
+    implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8")
+    implementation("io.quarkus:quarkus-arc")
+    implementation("io.quarkus:quarkus-rest")
+    testImplementation("io.quarkus:quarkus-junit5")
+    testImplementation("io.rest-assured:rest-assured")
+}
+
+group = "org.acme"
+version = "1.0-SNAPSHOT"
+
+tasks.withType<Test> {
+    systemProperty("java.util.logging.manager", "org.jboss.logmanager.LogManager")
+}
+allOpen {
+    annotation("jakarta.ws.rs.Path")
+    annotation("jakarta.enterprise.context.ApplicationScoped")
+    annotation("jakarta.persistence.Entity")
+    annotation("io.quarkus.test.junit.QuarkusTest")
+}
+
+kotlin {
+    compilerOptions {
+        javaParameters = true
+    }
+}

--- a/integration-tests/gradle/src/main/resources/included-kotlin-build/acme-subproject-nested/acme-nested/gradle.properties
+++ b/integration-tests/gradle/src/main/resources/included-kotlin-build/acme-subproject-nested/acme-nested/gradle.properties
@@ -1,0 +1,8 @@
+#Quarkus Gradle TS
+#Tue May 06 21:17:59 CEST 2025
+kotlinVersion=${kotlin.version}
+quarkusPluginVersion=${project.version}
+quarkusPlatformArtifactId=quarkus-bom
+quarkusPluginId=io.quarkus
+quarkusPlatformGroupId=io.quarkus
+quarkusPlatformVersion=${project.version}

--- a/integration-tests/gradle/src/main/resources/included-kotlin-build/acme-subproject-nested/acme-nested/settings.gradle.kts
+++ b/integration-tests/gradle/src/main/resources/included-kotlin-build/acme-subproject-nested/acme-nested/settings.gradle.kts
@@ -1,0 +1,20 @@
+pluginManagement {
+    val quarkusPluginVersion: String by settings
+    val quarkusPluginId: String by settings
+    val kotlinVersion: String by settings
+    repositories {
+        mavenLocal {
+            content {
+                includeGroupByRegex("io.quarkus.*")
+                includeGroup("org.hibernate.orm")
+            }
+        }
+        mavenCentral()
+        gradlePluginPortal()
+    }
+    plugins {
+        id(quarkusPluginId) version quarkusPluginVersion
+        kotlin("jvm") version kotlinVersion
+        kotlin("plugin.allopen") version kotlinVersion
+    }
+}

--- a/integration-tests/gradle/src/main/resources/included-kotlin-build/acme-subproject-nested/acme-nested/src/main/kotlin/org/example/SomeBean.kt
+++ b/integration-tests/gradle/src/main/resources/included-kotlin-build/acme-subproject-nested/acme-nested/src/main/kotlin/org/example/SomeBean.kt
@@ -1,0 +1,9 @@
+package org.example
+
+import jakarta.enterprise.context.ApplicationScoped
+
+@ApplicationScoped
+class SomeBean {
+
+    fun someMethod() = "Hello from SomeBean!"
+}

--- a/integration-tests/gradle/src/main/resources/included-kotlin-build/acme-subproject-nested/acme-nested/src/main/kotlin/org/example/SomeDataClass.kt
+++ b/integration-tests/gradle/src/main/resources/included-kotlin-build/acme-subproject-nested/acme-nested/src/main/kotlin/org/example/SomeDataClass.kt
@@ -1,0 +1,3 @@
+package org.example
+
+data class SomeDataClass(val someField: String)

--- a/integration-tests/gradle/src/main/resources/included-kotlin-build/acme-subproject-nested/acme-nested/src/main/resources/META-INF/beans.xml
+++ b/integration-tests/gradle/src/main/resources/included-kotlin-build/acme-subproject-nested/acme-nested/src/main/resources/META-INF/beans.xml
@@ -1,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns="http://xmlns.jcp.org/xml/ns/javaee"
+       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/beans_2_0.xsd"
+       bean-discovery-mode="annotated">
+</beans>

--- a/integration-tests/gradle/src/main/resources/included-kotlin-build/acme-subproject-nested/build.gradle.kts
+++ b/integration-tests/gradle/src/main/resources/included-kotlin-build/acme-subproject-nested/build.gradle.kts
@@ -1,0 +1,22 @@
+plugins {
+    kotlin("jvm")
+}
+
+repositories {
+    mavenLocal {
+        content {
+            includeGroupByRegex("io.quarkus.*")
+            includeGroup("org.hibernate.orm")
+        }
+    }
+    mavenCentral()
+}
+
+group = "org.acme"
+version = "1.0-SNAPSHOT"
+
+kotlin {
+    compilerOptions {
+        javaParameters = true
+    }
+}

--- a/integration-tests/gradle/src/main/resources/included-kotlin-build/acme-subproject-nested/gradle.properties
+++ b/integration-tests/gradle/src/main/resources/included-kotlin-build/acme-subproject-nested/gradle.properties
@@ -1,0 +1,8 @@
+#Quarkus Gradle TS
+#Tue May 06 21:43:16 CEST 2025
+kotlinVersion=${kotlin.version}
+quarkusPluginVersion=${project.version}
+quarkusPlatformArtifactId=quarkus-bom
+quarkusPluginId=io.quarkus
+quarkusPlatformGroupId=io.quarkus
+quarkusPlatformVersion=${project.version}

--- a/integration-tests/gradle/src/main/resources/included-kotlin-build/acme-subproject-nested/settings.gradle.kts
+++ b/integration-tests/gradle/src/main/resources/included-kotlin-build/acme-subproject-nested/settings.gradle.kts
@@ -1,0 +1,24 @@
+pluginManagement {
+    val quarkusPluginVersion: String by settings
+    val quarkusPluginId: String by settings
+    val kotlinVersion: String by settings
+    repositories {
+        mavenLocal {
+            content {
+                includeGroupByRegex("io.quarkus.*")
+                includeGroup("org.hibernate.orm")
+            }
+        }
+        mavenCentral()
+        gradlePluginPortal()
+    }
+    plugins {
+        id(quarkusPluginId) version quarkusPluginVersion
+        kotlin("jvm") version kotlinVersion
+        kotlin("plugin.allopen") version kotlinVersion
+    }
+}
+
+rootProject.name = "acme-subproject-nested"
+
+include("acme-nested")

--- a/integration-tests/gradle/src/main/resources/included-kotlin-build/build.gradle.kts
+++ b/integration-tests/gradle/src/main/resources/included-kotlin-build/build.gradle.kts
@@ -1,0 +1,51 @@
+plugins {
+    kotlin("jvm")
+    kotlin("plugin.allopen")
+    id("io.quarkus")
+}
+
+repositories {
+    mavenLocal {
+        content {
+            includeGroupByRegex("io.quarkus.*")
+            includeGroup("org.hibernate.orm")
+        }
+    }
+    mavenCentral()
+}
+
+val quarkusPlatformGroupId: String by project
+val quarkusPlatformArtifactId: String by project
+val quarkusPlatformVersion: String by project
+
+dependencies {
+    implementation("org.acme:acme-nested")
+
+
+    implementation(enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}"))
+    implementation("io.quarkus:quarkus-kotlin")
+    implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8")
+    implementation("io.quarkus:quarkus-arc")
+    implementation("io.quarkus:quarkus-rest")
+    testImplementation("io.quarkus:quarkus-junit5")
+    testImplementation("io.rest-assured:rest-assured")
+}
+
+group = "org.acme"
+version = "1.0-SNAPSHOT"
+
+tasks.withType<Test> {
+    systemProperty("java.util.logging.manager", "org.jboss.logmanager.LogManager")
+}
+allOpen {
+    annotation("jakarta.ws.rs.Path")
+    annotation("jakarta.enterprise.context.ApplicationScoped")
+    annotation("jakarta.persistence.Entity")
+    annotation("io.quarkus.test.junit.QuarkusTest")
+}
+
+kotlin {
+    compilerOptions {
+        javaParameters = true
+    }
+}

--- a/integration-tests/gradle/src/main/resources/included-kotlin-build/gradle.properties
+++ b/integration-tests/gradle/src/main/resources/included-kotlin-build/gradle.properties
@@ -1,0 +1,5 @@
+# Gradle properties
+quarkusPluginId=io.quarkus
+quarkusPlatformGroupId=io.quarkus
+quarkusPlatformArtifactId=quarkus-bom
+kotlinVersion=${kotlin.version}

--- a/integration-tests/gradle/src/main/resources/included-kotlin-build/settings.gradle.kts
+++ b/integration-tests/gradle/src/main/resources/included-kotlin-build/settings.gradle.kts
@@ -1,0 +1,23 @@
+pluginManagement {
+    val quarkusPluginVersion: String by settings
+    val quarkusPluginId: String by settings
+    val kotlinVersion: String by settings
+    repositories {
+        mavenLocal {
+            content {
+                includeGroupByRegex("io.quarkus.*")
+                includeGroup("org.hibernate.orm")
+            }
+        }
+        mavenCentral()
+        gradlePluginPortal()
+    }
+    plugins {
+        id(quarkusPluginId) version quarkusPluginVersion
+        kotlin("jvm") version kotlinVersion
+        kotlin("plugin.allopen") version kotlinVersion
+    }
+}
+rootProject.name = "acme-root-kotlin"
+
+includeBuild("acme-subproject-nested")

--- a/integration-tests/gradle/src/main/resources/included-kotlin-build/src/main/kotlin/org/example/ExampleResource.kt
+++ b/integration-tests/gradle/src/main/resources/included-kotlin-build/src/main/kotlin/org/example/ExampleResource.kt
@@ -1,0 +1,14 @@
+package org.example
+
+import jakarta.ws.rs.GET
+import jakarta.ws.rs.Path
+import jakarta.ws.rs.Produces
+import jakarta.ws.rs.core.MediaType
+
+@Path("/hello")
+class ExampleResource(val someBean: SomeBean) {
+
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    fun hello() = someBean.someMethod()
+}

--- a/integration-tests/gradle/src/test/java/io/quarkus/gradle/devmode/IncludedKotlinBuildDevModeTest.java
+++ b/integration-tests/gradle/src/test/java/io/quarkus/gradle/devmode/IncludedKotlinBuildDevModeTest.java
@@ -1,0 +1,23 @@
+package io.quarkus.gradle.devmode;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.Map;
+
+public class IncludedKotlinBuildDevModeTest extends QuarkusDevGradleTestBase {
+
+    @Override
+    protected String projectDirectoryName() {
+        return "included-kotlin-build";
+    }
+
+    @Override
+    protected void testDevMode() throws Exception {
+        assertThat(getHttpResponse("/hello")).contains("Hello from SomeBean!");
+
+        replace("acme-subproject-nested/acme-nested/src/main/kotlin/org/example/SomeBean.kt",
+                Map.of("Hello from SomeBean!", "Bye!"));
+
+        assertUpdatedResponseContains("/hello", "Bye!");
+    }
+}

